### PR TITLE
feat(validation): warn about NoInterfaceObject

### LINKED ISSUE #372: Add severity field for validations
We might warn about future removal of `[NoInterfaceObject]` but should *not* throw because the spec still has it.

### PR DESCRIPTION
**Pull request by saschanaz**
Closes #372

### CODE REVIEW

File: test/invalid/baseline/exposed.txt [hunk 1]
```diff
@@ -1,6 +1,9 @@
 Validation error at line 1 in exposed.webidl, inside `interface UnexposedInterface`:
 interface UnexposedInterface {};
           ^ Interfaces must have `[Exposed]` extended attribute. To fix, add, for example, `[Exposed=Window]`. Please also consider carefully if your interface should also be exposed in a Worker scope. Refer to the [WebIDL spec section on Exposed](https://heycam.github.io/webidl/#Exposed) for more information.
+Validation error at line 3 in exposed.webidl, inside `extended-attribute NoInterfaceObject`, with severity "weak":
```
Comment (marcoscaceres) [inline test/invalid/baseline/exposed.txt]:
```suggestion
Validation error at line 3 in exposed.webidl, inside `extended-attribute NoInterfaceObject`:
```

File: lib/error.js [hunk 2]
```diff
@@ -7,10 +7,14 @@ function lastLine(text) {
...
 /**
+ * @typedef {object} WebIDL2ErrorOptions
+ * @property {"normal" | "weak"} severity
+ *
  * @param {string} message error message
  * @param {"Syntax" | "Validation"} type error type
+ * @param {WebIDL2ErrorOptions} [options]
  */
-function error(source, position, current, message, type) {
+function error(source, position, current, message, type, { severity = "normal" } = {}) {
```
[the remark this answers is not in the document]
Comment (saschanaz) [reply lib/error.js]: `severity = info`? I think it's not really severity then... WDYT?

[the remark this answers is not in the document]
Comment (saschanaz) [reply lib/error.js]: Not sure about the warning/error either. Maybe `type=warning`?

[the remark this answers is not in the document]
Comment (marcoscaceres) [reply lib/error.js]: yeah, type is good.

[the remark this answers is not in the document]
Comment (marcoscaceres) [reply lib/error.js]: Or logLevel ?

[the remark this answers is not in the document]
Comment (marcoscaceres) [reply lib/error.js]: ... actually, `type` seems best...

[the remark this answers is not in the document]
Comment (saschanaz) [reply lib/error.js]: > ... actually, `type` seems best...

Thought that `logLevel` or `level` seems great as there are other libraries or frameworks using that word. `type` is IMO a bit too easy to abuse everywhere 😁

[the remark this answers is not in the document]
Comment (saschanaz) [reply lib/error.js]: Just renamed `type` as `kind` and `severity` as `level`, with intention to change to `type` again if we have a good reason...

File: lib/productions/extended-attributes.js [hunk 3]
```diff
@@ -75,6 +77,13 @@ class SimpleExtendedAttribute extends Base {
   }
 
   *validate(defs) {
+    if (this.name === "NoInterfaceObject") {
+      const message = `\`[NoInterfaceObject]\` extended attribute is an \
+undesirable feature that can be removed in the future. Refer to the \
```
Comment (marcoscaceres) [inline lib/productions/extended-attributes.js]:
```suggestion
undesirable feature that may be removed from WebIDL in the future. Refer to the \
```

File: test/invalid/baseline/exposed.txt [hunk 4]
```diff
@@ -1,6 +1,9 @@
 Validation error at line 1 in exposed.webidl, inside `interface UnexposedInterface`:
 interface UnexposedInterface {};
           ^ Interfaces must have `[Exposed]` extended attribute. To fix, add, for example, `[Exposed=Window]`. Please also consider carefully if your interface should also be exposed in a Worker scope. Refer to the [WebIDL spec section on Exposed](https://heycam.github.io/webidl/#Exposed) for more information.
+Validation error at line 3 in exposed.webidl, inside `extended-attribute NoInterfaceObject`:
+[NoInterfaceObject]
+ ^ `[NoInterfaceObject]` extended attribute is an undesirable feature that can be removed in the future. Refer to the [relevant upstream PR](https://github.com/heycam/webidl/pull/609) for more information.
```
Comment (marcoscaceres) [inline test/invalid/baseline/exposed.txt]:
```suggestion
 ^ `[NoInterfaceObject]` extended attribute is an undesirable feature that may be removed from WebIDL in the future. Refer to the [relevant upstream PR](https://github.com/heycam/webidl/pull/609) for more information.
```

File: lib/error.js [hunk 5]
```diff
@@ -7,10 +7,14 @@ function lastLine(text) {
 }
 
 /**
+ * @typedef {object} WebIDL2ErrorOptions
+ * @property {"normal" | "weak"} level
```
Comment (marcoscaceres) [inline lib/error.js]: Should this be "error" and "warning"?

Comment (saschanaz) [reply lib/error.js]: Oops, you're right.

```suggestion
 * @property {"error" | "warning"} level
```

File: lib/productions/extended-attributes.js (hunk 3 above)
Comment (saschanaz) [reply lib/productions/extended-attributes.js]: Maybe "Web IDL" because that's the official title?

File: test/invalid/baseline/exposed.txt (hunk 4 above)
Comment (marcoscaceres) [reply test/invalid/baseline/exposed.txt]: ok fixed both.

File: lib/productions/extended-attributes.js (hunk 3 above)
Comment (marcoscaceres) [reply lib/productions/extended-attributes.js]: Done

**Comment by saschanaz** [comment]
> Approving as only nits left :)

Are those nits removed now?

**Comment by marcoscaceres** [comment]
Yep, everything looks good.

**Pull request merged**